File: generated_code/cpp/test_unittest_4007.cpp
#include <gtest/gtest.h>
#include <vector>
#include "solution_proxy.h"

using namespace std;

// Test Case 1: Covering the provided example scenario
TEST(MaxTotalValueTest, ExampleCase) {
    Solution sol;
    vector<int> nums = {1, 3, 2};
    int k = 2;
    // Subarrays chosen: nums[0..1] -> [1, 3] (val: 3-1=2) and nums[0..2] -> [1, 3, 2] (val: 3-1=2)
    // Total value = 2 + 2 = 4
    long long expected = 4;
    EXPECT_EQ(sol.maxTotalValue(nums, k), expected);
}

// Test Case 2: Minimal array size and minimum k
TEST(MaxTotalValueTest, MinimalInputEdgeCase) {
    Solution sol;
    vector<int> nums = {5};
    int k = 1;
    // Only one possible subarray: [5]. Value: 5 - 5 = 0
    long long expected = 0;
    EXPECT_EQ(sol.maxTotalValue(nums, k), expected);
}

// Test Case 3: All elements are identical
TEST(MaxTotalValueTest, IdenticalElements) {
    Solution sol;
    vector<int> nums = {7, 7, 7, 7};
    int k = 3;
    // Any subarray will have max == min, so value is always 0
    long long expected = 0;
    EXPECT_EQ(sol.maxTotalValue(nums, k), expected);
}

// Test Case 4: Strictly increasing array
TEST(MaxTotalValueTest, StrictlyIncreasingArray) {
    Solution sol;
    vector<int> nums = {1, 2, 4, 7};
    int k = 3;
    // Potential subarrays and their values:
    // [1, 2, 4, 7] -> 7 - 1 = 6
    // [2, 4, 7]    -> 7 - 2 = 5
    // [1, 2, 4]    -> 4 - 1 = 3
    // [4, 7]       -> 7 - 4 = 3
    // Top 3 distinct subarrays with max values would be 6 + 5 + 3 = 14 (or 6 + 5 + 3 from another combination)
    long long expected = 14;
    EXPECT_EQ(sol.maxTotalValue(nums, k), expected);
}

// Test Case 5: Large values that could cause 32-bit integer overflow in total sum
TEST(MaxTotalValueTest, PotentialOverflowHandling) {
    Solution sol;
    vector<int> nums = {-1000000000, 1000000000, 1000000000, -1000000000};
    int k = 2;
    // Subarray 1: [ -1e9, 1e9 ] -> 1e9 - (-1e9) = 2,000,000,000
    // Subarray 2: [ 1e9, 1e9, -1e9 ] -> 1e9 - (-1e9) = 2,000,000,000
    // Total = 4,000,000,000 (Exceeds standard signed 32-bit int max of ~2.14x10^9)
    long long expected = 4000000000LL;
    EXPECT_EQ(sol.maxTotalValue(nums, k), expected);
}

// Test Case 6: Maxing out k to choose all possible subarrays
TEST(MaxTotalValueTest, SelectAllSubarrays) {
    Solution sol;
    vector<int> nums = {1, 5, 2};
    // Total non-empty subarrays for size 3 is n*(n+1)/2 = 6
    // Subarrays and values:
    // [1] -> 0, [5] -> 0, [2] -> 0
    // [1, 5] -> 4
    // [5, 2] -> 3
    // [1, 5, 2] -> 4
    // Total sum = 0 + 0 + 0 + 4 + 3 + 4 = 11
    int k = 6;
    long long expected = 11;
    EXPECT_EQ(sol.maxTotalValue(nums, k), expected);
}

File: generated_code/python/test_unittest_4007.py
import os
import importlib.util
import unittest

# Dynamic loading of the Solution class as mandated by the guidelines
solution_path = os.environ.get("TEST_SOLUTION_FILE")
if not solution_path:
    raise RuntimeError("Environment variable 'TEST_SOLUTION_FILE' is not set.")

spec = importlib.util.spec_from_file_location("Solution", solution_path)
sol_module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(sol_module)
Solution = sol_module.Solution

class TestMaxTotalValue(unittest.TestCase):
    def setUp(self):
        self.sol = Solution()

    def test_example_case(self):
        """Tests the example scenario described in the problem statement."""
        # Based on the problem text: nums = [1, 3, 2], k = 2
        # Chosen subarrays: nums[0..1] -> [1, 3] (val: 2), nums[0..2] -> [1, 3, 2] (val: 2)
        # Total value: 2 + 2 = 4
        nums = [1, 3, 2]
        k = 2
        expected = 4
        self.assertEqual(self.sol.maxTotalValue(nums, k), expected)

    def test_single_element_subarrays(self):
        """Tests when subarrays have a length of 1, yielding a value of 0."""
        nums = [5, 5, 5]
        k = 3
        # Any single element subarray [5] has max(5) - min(5) = 0
        expected = 0
        self.assertEqual(self.sol.maxTotalValue(nums, k), expected)

    def test_all_elements_same(self):
        """Tests an array where all elements are identical; any subarray value will be 0."""
        nums = [10, 10, 10, 10]
        k = 5
        expected = 0
        self.assertEqual(self.sol.maxTotalValue(nums, k), expected)

    def test_minimum_constraints(self):
        """Tests the absolute minimum bounds for the input constraints."""
        nums = [1]
        k = 1
        # Only one possible subarray: nums[0..0] -> [1], value = 1 - 1 = 0
        expected = 0
        self.assertEqual(self.sol.maxTotalValue(nums, k), expected)

    def test_strictly_increasing_array(self):
        """Tests a strictly increasing array where max and min are strictly determined by boundaries."""
        nums = [1, 2, 4, 7]
        k = 3
        # Highly valuable subarrays:
        # nums[0..3] -> [1, 2, 4, 7] (val: 7 - 1 = 6)
        # nums[1..3] -> [2, 4, 7]    (val: 7 - 2 = 5)
        # nums[0..2] -> [1, 2, 4]    (val: 4 - 1 = 3)
        # Total: 6 + 5 + 3 = 14
        nums = [1, 2, 4, 7]
        k = 3
        expected = 14
        self.assertEqual(self.sol.maxTotalValue(nums, k), expected)

    def test_strictly_decreasing_array(self):
        """Tests a strictly decreasing array where max and min are reversed in terms of indices."""
        nums = [10, 7, 4, 1]
        k = 2
        # Best 2 subarrays:
        # nums[0..3] -> [10, 7, 4, 1] (val: 10 - 1 = 9)
        # nums[0..2] -> [10, 7, 4]    (val: 10 - 4 = 6) or nums[1..3] -> [7, 4, 1] (val: 7 - 1 = 6)
        # Total: 9 + 6 = 15
        nums = [10, 7, 4, 1]
        k = 2
        expected = 15
        self.assertEqual(self.sol.maxTotalValue(nums, k), expected)

    def test_negative_and_positive_mixed(self):
        """Tests an array containing both negative and positive integers."""
        nums = [-5, 10, -2, 6]
        k = 2
        # Best 2 subarrays:
        # nums[0..3] -> [-5, 10, -2, 6] (val: 10 - (-5) = 15)
        # nums[0..1] -> [-5, 10]         (val: 10 - (-5) = 15)
        # Total: 15 + 15 = 30
        nums = [-5, 10, -2, 6]
        k = 2
        expected = 30
        self.assertEqual(self.sol.maxTotalValue(nums, k), expected)

    def test_large_k_choosing_all_possible_subarrays(self):
        """Tests when k is equal to the total number of non-empty subarrays possible."""
        nums = [1, 5, 2]
        # Total non-empty subarrays = (3 * 4) // 2 = 6
        # Subarrays and values:
        # [1] -> 0, [5] -> 0, [2] -> 0
        # [1, 5] -> 4, [5, 2] -> 3
        # [1, 5, 2] -> 4
        # Total sum of all values = 0 + 0 + 0 + 4 + 3 + 4 = 11
        k = 6
        expected = 11
        self.assertEqual(self.sol.maxTotalValue(nums, k), expected)

if __name__ == "__main__":
    unittest.main()